### PR TITLE
Save and retrieve local history under the user's home

### PR DESCRIPTION
**Pull request by Anasshahidd21**
Fixes: https://github.com/vince-fugnitto/local-history-ext/issues/37

**What it does**
Creates a .local-history folder in the home directory of the user,
it then also hashes the file uri and creates folder on each uri.
Currently it is stored in each workspace.

**How to Test**
1. Add a new folder to the workspace
2. Save something on an active editor 
4. Go to your local home directory
5. navigate to .local-history
6. Check to see if there exists a folder name with hashcode
7. Open the folder to see if there are revisions.
8. Open the Command Palette (Ctrl+Shift+P), enter Local History: Local History: Active Editor
9. Right click in the active editor to trigger the editor's context menu and select Local History: Active Editor
10. Select a local history file to view the diff
11. Click the `Trash` icon in the top right menu when in diff mode to delete the revision
12. Click the `Revert` icon in the top right menu when in diff mode to revert to previous revision

Signed-off-by: Anas Shahid <muhammad.shahid@ericsson.com>

### CODE REVIEW

File: src/local-history/local-history-manager.ts [hunk 1]
```diff
@@ -2,61 +2,58 @@ import * as fs from 'fs';
...
+     * @param encoding: the required conversion of data received by the stream.
+     */
+    private readFile(uri: string, encoding = 'utf8'): Promise<string> {
+        const readStream = fs.createReadStream(uri, { encoding });
+        return new Promise((resolve) => {
+            let data = '';
+            readStream.on('data', chunk => data += chunk);
+            readStream.on('end', () => resolve(data));
+        });
+    }
```
Comment (Anasshahidd21) [inline src/local-history/local-history-manager.ts:44]: we can revert it back, as it is not part of this PR

File: src/local-history/local-history-manager.ts [hunk 2]
```diff
@@ -2,61 +2,58 @@ import * as fs from 'fs';
...
+     * @param uri: the target uri for the new file.
+     * @param content: the data to be stored in the file
+     */
+    private writeFile(uri: string, content: string): Promise<string> {
+        const writeStream = fs.createWriteStream(uri);
+        return new Promise((resolve) => {
+            writeStream.on('open', () => writeStream.write(content));
+            writeStream.on('end', () => resolve());
+        });
     }
```
Comment (Anasshahidd21) [inline src/local-history/local-history-manager.ts:57]: we can revert it back, as it is not part of this PR

File: src/local-history/local-history-manager.ts [hunk 3]
```diff
@@ -75,93 +72,117 @@ export class LocalHistoryManager {
...
+                        timestamp: this.parseTimestamp(fileName),
+                        uri: path.join(hashedFolderPath, fileName),
+                    };
+                    this.historyFilesForActiveEditor.unshift(data);
+                }
+            }
+        } catch (err) {
+            vscode.window.showInformationMessage(err);
+        }
+    }
```
Comment (Anasshahidd21) [inline src/local-history/local-history-manager.ts:110]: [question]: Should we display the same error to the user, or handle it more gracefully by creating an error string specific to the case?

File: src/local-history/local-history-manager.ts [hunk 4]
```diff
@@ -75,93 +72,117 @@ export class LocalHistoryManager {
...
 
-            // Copy the content of the current active editor.
-            const historyFilePath = path.join(workspaceFolderPath, '.local-history', historyFileName);
-            const data: HistoryFileProperties = {
-                fileName: historyFileName,
-                timestamp: timestampForProperty,
-                uri: historyFilePath,
-                parentFileName: fileFullPath.base
-            };
-            this.historyFiles.unshift(data);
```
Comment (Anasshahidd21) [inline src/local-history/local-history-manager.ts:161]: historyFiles should be updated with the content.

File: src/local-history/local-history-manager.ts [hunk 5]
```diff
@@ -75,93 +72,117 @@ export class LocalHistoryManager {
...
      */
-    public viewAllForActiveEditor(textEditor: vscode.TextEditor): void {
+    public viewAllHistoryForActiveEditor(textEditor: vscode.TextEditor): void {
 
         if (vscode.workspace.getWorkspaceFolder(textEditor.document.uri) !== undefined) {
-            const workspaceFolderPath: string = vscode.workspace.getWorkspaceFolder(textEditor.document.uri)!.uri.fsPath;
-            const activeEditor = path.parse(textEditor.document.fileName);
+            const hashedEditorPath = checksum(textEditor.document.fileName);
+            const localHistoryFolderPath = path.join(os.homedir(), '.local-history');
+            const hashedFolderPath = path.join(localHistoryFolderPath, hashedEditorPath);
```
Comment (Anasshahidd21) [inline src/local-history/local-history-manager.ts:120]: we can change it to  `const hashedFolderPath = path.join(os.homedir(), '.local-history', hashedEditorPath);`

File: src/local-history/local-history-manager.ts [hunk 6]
```diff
@@ -75,93 +72,117 @@ export class LocalHistoryManager {
...
         const timestampForFileName = timestamp.replace(/[-:. ]/g, '');
-        const timestampForProperty = timestamp.substring(0, 19);
         const fileFullPath = path.parse(document.fileName);
         const historyFileName = `${fileFullPath.name}_${timestampForFileName.substring(0, 14)}_${timestampForFileName.substring(14, 17)}${fileFullPath.ext}`;
 
         if (vscode.workspace.getWorkspaceFolder(document.uri) !== undefined) {
-            const workspaceFolderPath: string = vscode.workspace.getWorkspaceFolder(document.uri)!.uri.fsPath;
+            const localHistoryFolderPath = path.join(os.homedir(), '.local-history');
+            const hashedEditorPath = checksum(document.fileName);
+            const hashedFolderPath = path.join(localHistoryFolderPath, hashedEditorPath);
```
Comment (Anasshahidd21) [inline src/local-history/local-history-manager.ts:171]: we can change it to  `const hashedFolderPath = path.join(os.homedir(), '.local-history', hashedEditorPath);`

File: src/local-history/local-history-manager.ts (hunk 4 above)
Comment (kaiyue0329) [reply src/local-history/local-history-manager.ts:161]: Instead of all the local history file, `historyFiles` now stores the history files for the active editor. We load history files from user's home each time we trigger `View All for Active Editor`.

**Comment by vince-fugnitto** [comment]
Ping me whenever it's ready for review :)

File: src/extension.ts [hunk 7]
```diff
@@ -35,7 +35,7 @@ export function activate(context: vscode.ExtensionContext) {
     disposable.push(
         vscode.commands.registerTextEditorCommand('local-history.removeRevision', () => {
             const editors = vscode.window.visibleTextEditors;
-            if (editors && editors.length > 0) {
+            if (editors && editors.length === 2) {
```
Comment (vince-fugnitto) [inline src/extension.ts:38]: Do you mind clarifying why we are checking if the length is exactly 2?

File: src/local-history/local-history-manager.ts [hunk 8]
```diff
@@ -2,60 +2,58 @@ import * as fs from 'fs';
...
-                const parentFileName = fileName.match('^[^_/]*')![0] + fileName.match('.[0-9a-z]+$')![0];
-                const data: HistoryFileProperties = {
-                    fileName: fileName,
-                    timestamp: this.parseTimestamp(fileName),
-                    uri: path.join(folderUri.path, fileName),
-                    parentFileName: parentFileName
-                };
-                this.historyFiles.unshift(data);
+        this.historyFilesForActiveEditor = [];
+        const pathInVscodeUri = vscode.Uri.file(hashedFolderPath);
```
Comment (vince-fugnitto) [inline src/local-history/local-history-manager.ts:42]: Can we have a better name for this variable?
Just by reading it it is not clear as to what it actually is.

File: src/local-history/local-history-manager.ts [hunk 9]
```diff
@@ -2,60 +2,58 @@ import * as fs from 'fs';
...
+                    const data: HistoryFileProperties = {
+                        fileName: fileName,
+                        timestamp: this.parseTimestamp(fileName),
+                        uri: path.join(hashedFolderPath, fileName),
+                    };
+                    this.historyFilesForActiveEditor.unshift(data);
+                }
             }
+        } catch (err) {
+            console.warn('An error occurs when reading the .local-history directory', err);
```
Comment (vince-fugnitto) [inline src/local-history/local-history-manager.ts:56]: Past tense would make more sense:

```
An error has occurred when reading the .local-history directory.
```

File: src/local-history/local-history-manager.ts [hunk 10]
```diff
@@ -78,43 +76,43 @@ export class LocalHistoryManager {
     /**
      * View all history of the active editor.
      */
-    public viewAllForActiveEditor(textEditor: vscode.TextEditor): void {
+    public viewAllHistoryForActiveEditor(textEditor: vscode.TextEditor): void {
```
Comment (vince-fugnitto) [inline src/local-history/local-history-manager.ts:79]: I don't really like the naming of this method, or the command itself.
I think we can be more concise with naming variables, ex:

```ts
/**
* View the local-history for the given editor.
*/
viewHistory(editor: vscode.TextEditor): void { }
```

File: src/local-history/local-history-manager.ts [hunk 11]
```diff
@@ -219,25 +215,19 @@ export class LocalHistoryManager {
...
-                        if (err && err.code === 'ENOENT') {
-                            vscode.window.showInformationMessage('File is not a revision.');
-                            return;
-                        }
-                        else if (err) {
-                            vscode.window.showInformationMessage(err.message);
+                    // Remove the revision.
+                    fs.unlink(revision.document.fileName, (err) => {
+                        if (err) {
+                            console.warn('An error occurs when removing the revision', err.message);
```
Comment (vince-fugnitto) [inline src/local-history/local-history-manager.ts:222]: Same as above.

File: src/local-history/local-history-manager.ts [hunk 12]
```diff
@@ -180,8 +172,12 @@ export class LocalHistoryManager {
      */
     public async revertActiveEditorToPrevRevision(previous: vscode.TextEditor, current: vscode.TextEditor): Promise<void> {
         if (current && previous) {
-            const fileContent = await this.readFile(previous.document.fileName);
-            await this.writeFile(current.document.fileName, fileContent);
+            try {
+                const fileContent = await this.readFile(previous.document.fileName);
+                await this.writeFile(current.document.fileName, fileContent);
+            } catch (err) {
+                console.warn('An error occurs when reverting to previous revision', err);
```
Comment (vince-fugnitto) [inline src/local-history/local-history-manager.ts:179]: Same as above.

File: src/local-history/local-history-manager.ts [hunk 13]
```diff
@@ -136,32 +134,26 @@ export class LocalHistoryManager {
...
-
-            const activeDocumentContent: string = await this.readFile(document.fileName);
-            await this.writeFile(historyFilePath, activeDocumentContent);
+            try {
+                const historyFilePath = path.join(hashedFolderPath, historyFileName);
+                // Copy the content of the current active editor.
+                const activeDocumentContent: string = await this.readFile(document.fileName);
+                await this.writeFile(historyFilePath, activeDocumentContent);
+            } catch (err) {
+                console.warn('An error occurs when saving the active editor content', err);
```
Comment (vince-fugnitto) [inline src/local-history/local-history-manager.ts:155]: Same as above.

File: src/local-history/local-history-manager.ts [hunk 14]
```diff
@@ -78,43 +76,43 @@ export class LocalHistoryManager {
...
+                        matchOnDescription: true,
+                        matchOnDetail: true
+                    }).then((selection) => {
+                        // User made final selection.
+                        if (!selection) {
+                            return;
+                        }
+
+                        // Get the file system path for the selection.
+                        const selectionFsPath = path.join(hashedFolderPath, selection.description!);
```
Comment (vince-fugnitto) [inline src/local-history/local-history-manager.ts:111]: These changes look unrelated? Is it perhaps due to formatting issues?

File: src/extension.ts (hunk 7 above)
Comment (kaiyue0329) [reply src/extension.ts:38]: length is exactly 2 when we are in diff mode since we have two editors visible side-by-side.

Comment (vince-fugnitto) [reply src/extension.ts:38]: I’m not sure we should do such a check, if we were to close the original file and keep only the diff would it still hold true? Or if we had multiple editors opened.

File: src/local-history/local-history-manager.ts (hunk 8 above)
Comment (kaiyue0329) [reply src/local-history/local-history-manager.ts:42]: It converts a path (string) into a vscode.uri. Would `hashFolderUri` be a better name for this variable?

Comment (vince-fugnitto) [reply src/local-history/local-history-manager.ts:42]: I think that would make more sense, it’s just the URI, no real need to call it a vscodeURI

File: src/extension.ts (hunk 7 above)
Comment (kaiyue0329) [reply src/extension.ts:38]: I think this check would cause this issue https://github.com/vince-fugnitto/local-history-ext/issues/40. The issue no longer exists when I will change it back to `length > 0`.

File: src/local-history/local-history-manager.ts (hunk 14 above)
Comment (kaiyue0329) [reply src/local-history/local-history-manager.ts:111]: Yes.

File: src/local-history/local-history-manager.ts [hunk 15]
```diff
@@ -136,32 +134,26 @@ export class LocalHistoryManager {
     public async saveActiveEditorContext(document: vscode.TextDocument): Promise<void> {
```
Comment (kaiyue0329) [inline src/local-history/local-history-manager.ts:134]: @vince-fugnitto Should I change this method name to `saveEditorContext` so that it is more concise?

File: src/extension.ts (hunk 7 above)
Comment (kaiyue0329) [reply src/extension.ts:38]: It still hold true when we kept only the diff or had multiple editors opened. I was still able to delete the revision.

File: src/local-history/local-history-manager.ts [hunk 16]
```diff
@@ -78,43 +76,43 @@ export class LocalHistoryManager {
...
-                    description: vscode.workspace.asRelativePath(item.uri)
+                    description: path.basename(item.uri)
                 }));
 
-            if (items.length === 0) {
-                vscode.window.showInformationMessage(`No local history file found for '${activeEditor.base}'`);
-                return;
-            }
+                if (items.length === 0) {
+                    vscode.window.showInformationMessage(`No local history file found for '${path.basename(editor.document.fileName)}'`);
```
Comment (vince-fugnitto) [inline src/local-history/local-history-manager.ts:92]: [minor] 'No local history found for file...'
